pkg/logql/range_vector.go

```diff
@@ -456,6 +488,51 @@ func quantileOverTime(q float64) func(samples []promql.Point) float64 {
 	}
 }
 
+func bucketsOverTime(nativeHistogramBucketFactor float64) HistogramVectorAggregator {
+	return func(samples []promql.Point) *histogram.FloatHistogram {
+		sampleHistogram := client_golang.NewHistogram(client_golang.HistogramOpts{
+			NativeHistogramBucketFactor: nativeHistogramBucketFactor,
+			Name:                        "logql_buckets", // not sure if this should be hard coded
+		})
+		for _, sample := range samples {
+			sampleHistogram.Observe(sample.V)
+		}
+		floatHistogram := &dto.Metric{}
+		// this might be a costly operation for large values, at the moment,
+		// this is the only way I could find to convert to floatHistogram
+		if err := sampleHistogram.Write(floatHistogram); err != nil {
+			// should add handled with proper logging
+			fmt.Printf("err in hist=%s", err.Error())
+			return nil
+		}
+		positiveSpans := make([]histogram.Span, len(floatHistogram.Histogram.GetPositiveSpan()))
+
+		for i, span := range floatHistogram.Histogram.GetPositiveSpan() {
+			positiveSpans[i].Length = *span.Length
+			positiveSpans[i].Offset = *span.Offset
+		}
+		negativeSpans := make([]histogram.Span, len(floatHistogram.Histogram.GetNegativeSpan()))
+
+		for i, span := range floatHistogram.Histogram.GetNegativeSpan() {
+			negativeSpans[i].Length = *span.Length
+			negativeSpans[i].Offset = *span.Offset
+		}
+
+		h := &histogram.Histogram{
+			Schema:          floatHistogram.Histogram.GetSchema(),
+			ZeroThreshold:   floatHistogram.Histogram.GetZeroThreshold(),
+			ZeroCount:       floatHistogram.Histogram.GetZeroCount(),
+			Count:           floatHistogram.Histogram.GetSampleCount(),
+			Sum:             floatHistogram.Histogram.GetSampleSum(),
+			PositiveSpans:   positiveSpans,
+			NegativeSpans:   negativeSpans,
+			PositiveBuckets: floatHistogram.Histogram.GetPositiveDelta(),
+			NegativeBuckets: floatHistogram.Histogram.GetNegativeDelta(),
+		}
+		return h.ToFloat()
```

Comment:
What's the reason for going through all this transformation work to use a `FloatHistogram` specifically?

Reply:
`promql.Point` only supports `FloatHistogram` https://github.com/prometheus/prometheus/blob/main/promql/value.go#L110

Reply:
Ah, looks like sparse histograms actually changed the data model itself in prometheus responses (see `note`). I don't quite understand it yet.
<img width="872" alt="image" src="https://user-images.githubusercontent.com/8173478/217254724-29964bbd-0ea6-4a8c-ad75-90336cde4930.png">